File: backend/services/post-ktor-mongo-fdm/src/main/kotlin/me/jason5lee/post_ktor_mongo_fdm/create_post/impl.kt
package me.jason5lee.post_ktor_mongo_fdm.create_post

import com.mongodb.MongoWriteException
import kotlinx.coroutines.reactive.awaitSingle
import me.jason5lee.post_ktor_mongo_fdm.common.PostContent
import me.jason5lee.post_ktor_mongo_fdm.common.PostId
import me.jason5lee.post_ktor_mongo_fdm.common.UserId
import me.jason5lee.post_ktor_mongo_fdm.common.api.invalidAuth
import me.jason5lee.post_ktor_mongo_fdm.common.utils.Db
import me.jason5lee.post_ktor_mongo_fdm.common.utils.Deps
import me.jason5lee.post_ktor_mongo_fdm.common.utils.timeFromNow
import org.bson.Document

class WorkflowImpl(val deps: Deps) : Workflow(), ErrorsImpl {
    override suspend fun run(caller: UserId, input: Command): PostId {
        val doc = Document()
        when (val content = input.content) {
            is PostContent.Text -> doc["text"] = content.value.value
            is PostContent.Url -> doc["url"] = content.value.value
        }
        doc["creator"] = Db.toObjectIdOrNull(caller.value) ?: throw invalidAuth()
        doc["creationTime"] = timeFromNow(offset = 0).utc
        doc["title"] = input.title.value

        val result = try {
            deps.mongoDb.getCollection(Db.posts).insertOne(doc).awaitSingle()
        } catch (e: MongoWriteException) { // I was also surprised that this is not DuplicateKeyException.
            throw duplicateTitle()
        }
        return PostId(Db.formatId(result.insertedId!!.asObjectId().value))
    }
}

File: backend/services/post-ts-koa-mongo/src/create-post/impl.ts
import { Deps, now } from "../common/utils";
import { Command, Workflow } from ".";
import { PostId, UserId } from "../common";
import * as db from "../common/utils/db";
import { Long, MongoError, ObjectId } from "mongodb";
import { errors } from "./api";
import { invalidAuth } from "../common/api/auth";

export class WorkflowImpl implements Workflow {
    async run(caller: UserId, input: Command): Promise<PostId> {
        let contentObj: { text?: string, url?: string };
        switch (input.content.type) {
            case "Text":
                contentObj = { text: input.content.content };
                break;
            case "Url":
                contentObj = { url: input.content.content };
                break;
        }
        const creator = db.tryParseId(caller);
        if (creator === undefined) {
            throw invalidAuth();
        }
        try {
            const post = await this.deps.mongoDb.collection(db.posts).insertOne({
                creator: creator,
                creationTime: Long.fromNumber(now().utc),
                title: input.title,
                ...contentObj,
            } satisfies {
                creator: ObjectId,
                creationTime: Long,
                title: string,
                text?: string,
                url?: string,
            });
            return db.formatId(post.insertedId) as PostId;
        } catch (e) {
            if (e instanceof MongoError) {
                if (e.code === 11000) {
                    throw this.errors.duplicateTitle();
                }
            }
            throw e;
        }
    }

    readonly errors = errors;
    constructor(private deps: Deps) { }
}
